### PR TITLE
feat(presences): #MA-899 add lateness motif

### PR DESCRIPTION
**Pull request by BastienBenard**

### CODE REVIEW

File: presences/src/main/java/fr/openent/presences/helper/EventQueryHelper.java [hunk 1]
```diff
@@ -160,53 +168,54 @@ public static String filterTimes(String startTime, String endTime, JsonArray par
...
      *
      * @param reasonIds   list of reason ids
-     * @param noReason    no reason wish to not display boolean
+     * @param noAbsenceReason    no reason wish to not display boolean
+     * @param noLatenessReason    no reason wish to not display boolean
      * @param regularized regularized wish to display boolean
      * @param params      JsonArray params where you add the reasonIds, no reason and regularized boolean
      * @return {String}
      */
-    public static String filterReasons(List<String> reasonIds, Boolean noReason, Boolean regularized,
```
Comment (BastienBenard) [inline presences/src/main/java/fr/openent/presences/helper/EventQueryHelper.java:168]: Nouveau:
`        if ((noReason != null && noReason) && (regularized != null && regularized)) {
            query += "AND (reason_id IS NULL OR (reason_id IS NOT NULL AND counsellor_regularisation = true)";
            if (reasonIds != null && !reasonIds.isEmpty()) {
                query += " AND reason_id IN " + Sql.listPrepared(reasonIds);
                params.addAll(new JsonArray(reasonIds));
            }
            query += ")";

            // else is default condition
        } else {
            // If we want to fetch events WITH reasonId, array reasonIds fetched is not empty
            // (optional if we wish noReason fetched at same time then noReason is TRUE)
            if (reasonIds != null && !reasonIds.isEmpty()) {
                query += " AND ((reason_id IN " + Sql.listPrepared(reasonIds) + ")";

                if (noReason != null && noReason) {
                    query += " OR reason_id IS NULL";
                } else {
                    query += typeIds.contains(EventType.LATENESS.getType().toString())
                            ? (" OR type_id = " + EventType.LATENESS.getType()) : "";
                }
                query += ")";
                params.addAll(new JsonArray(reasonIds));
            }

            // If we want to fetch events with NO reasonId, array reasonIds fetched is empty
            // AND noReason is TRUE
            if (reasonIds != null && reasonIds.isEmpty() && (noReason != null && noReason)) {
                query += " AND (reason_id IS NULL " + (regularized != null ? " OR counsellor_regularisation = " + regularized + "" : "") + ") ";
            }

            if (regularized != null) {
                query += " AND (counsellor_regularisation = " + regularized
                        + " OR type_id != " + EventType.ABSENCE.getType() +") ";
            }
        }`
Ancien:
`// this condition occurs when we want to filter no reason and regularized event at the same time
        if ((noReason != null && noReason) && (regularized != null && regularized)) {
            query += "AND (reason_id IS NULL OR (reason_id IS NOT NULL AND counsellor_regularisation = true)";
            if (reasonIds != null && !reasonIds.isEmpty()) {
                query += " AND reason_id IN " + Sql.listPrepared(reasonIds);
                params.addAll(new JsonArray(reasonIds));
            }
            query += ")";

            // else is default condition
        } else {
            // If we want to fetch events WITH reasonId, array reasonIds fetched is not empty
            // (optional if we wish noReason fetched at same time then noReason is TRUE)
            if (reasonIds != null && !reasonIds.isEmpty()) {
                query += " AND ((reason_id IN " + Sql.listPrepared(reasonIds) + ")";

                if (noReason != null && noReason) {
                    query += " OR reason_id IS NULL";
                } else {
                    query += typeIds.contains(EventType.LATENESS.getType().toString())
                            ? (" OR type_id = " + EventType.LATENESS.getType()) : "";
                }
                query += ")";
                params.addAll(new JsonArray(reasonIds));
            }

            // If we want to fetch events with NO reasonId, array reasonIds fetched is empty
            // AND noReason is TRUE
            if (reasonIds != null && reasonIds.isEmpty() && (noReason != null && noReason)) {
                query += " AND (reason_id IS NULL " + (regularized != null ? " OR counsellor_regularisation = " + regularized + "" : "") + ") ";
            }

            if (regularized != null) {
                query += " AND (counsellor_regularisation = " + regularized
                        + " OR type_id != " + EventType.ABSENCE.getType() +") ";
            }
        }`

File: presences/src/main/resources/public/sass/global/components/containers/events/_table.scss [hunk 2]
```diff
@@ -86,6 +86,16 @@
 }
 
 .table-reason {
+  &-icon {
+    position: absolute;
+    top: -3px;
+    left: 8px;
```
Comment (BastienBenard) [inline presences/src/main/resources/public/sass/global/components/containers/events/_table.scss:92]: image

Comment (JinDouang) [reply presences/src/main/resources/public/sass/global/components/containers/events/_table.scss:92]: 👍

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 3]
```diff
@@ -959,4 +985,12 @@ export const registersController = ng.controller('RegistersController',
                     window.filter = {};
                 }
             };
+
+            vm.getAbsenceReason = () => {
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: typage

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 4]
```diff
@@ -959,4 +985,12 @@ export const registersController = ng.controller('RegistersController',
...
+                return vm.reasons.filter((reason: Reason) => reason.reason_type_id == REASON_TYPE_ID.ABSENCE)
+            };
+
+            vm.getLatenessReason = () => {
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: typage

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 5]
```diff
@@ -713,6 +720,25 @@ export const registersController = ng.controller('RegistersController',
                 }
             };
 
+            vm.manageLateness = async function (events, student) {
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: typage

**Review by alicedraillard — APPROVED** [review]

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 6]
```diff
@@ -713,6 +720,25 @@ export const registersController = ng.controller('RegistersController',
...
+                if (events.id) {
+                    if (typeof events.register_id === 'string') {
+                        events.register_id = parseInt(events.register_id);
+                    }
+                    new Events().updateReason([events], events.reason_id, student.id, window.structure.id);
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: On a pas de service pour updateReason ?

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 7]
```diff
@@ -713,6 +720,25 @@ export const registersController = ng.controller('RegistersController',
...
+                } else {
+                    let reason_id = events.reason_id;
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: typage

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 8]
```diff
@@ -713,6 +720,25 @@ export const registersController = ng.controller('RegistersController',
...
+                    student.lateness = undefined;
+                    await vm.toggleLateness(student);
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: try catch ton await

File: presences/src/main/java/fr/openent/presences/service/impl/DefaultEventService.java [hunk 9]
```diff
@@ -708,9 +709,16 @@ public void changeReasonEvents(JsonObject eventBody, UserInfos user, Handler<Eit
...
-            editCorrespondingAbsences(events, user, eventBody.getString("student_id"), eventBody.getString("structure_id"),
-                    result.right().getValue().getJsonObject(0).getBoolean("counsellor_regularisation"),
-                    eventBody.getInteger("reasonId"), handler);
+            List<Event> absenceEvent = events.stream().filter(event -> event.getEventType() != null && EventType.ABSENCE.getType().equals(event.getEventType().getId()))
+                    .collect(Collectors.toList());
+            if (absenceEvent.isEmpty()) {
+                handler.handle(new Either.Right<>(new JsonObject().put("status", "ok")));
+            }
+            else {
+                editCorrespondingAbsences(absenceEvent, user, eventBody.getString("student_id"), eventBody.getString("structure_id"),
```
Comment (JinDouang) [inline presences/src/main/java/fr/openent/presences/service/impl/DefaultEventService.java]: attention magic string

**Comment by JinDouang** [comment]
motif = reason

Et..étant donné que tu as maj cette API, est-ce qu'il faut modifier la doc dans le contrat API et p-e notifier l'équiple appli mobile ?

File: presences/src/main/java/fr/openent/presences/service/impl/DefaultEventService.java [hunk 10]
```diff
@@ -708,9 +709,16 @@ public void changeReasonEvents(JsonObject eventBody, UserInfos user, Handler<Eit
...
+                        result.right().getValue().getJsonObject(0).getBoolean("counsellor_regularisation"),
```
Comment (JinDouang) [inline presences/src/main/java/fr/openent/presences/service/impl/DefaultEventService.java]: pas de risque sur ton result right get value qui est un tableau de jsonobject ?

File: presences/src/main/java/fr/openent/presences/helper/EventQueryHelper.java [hunk 11]
```diff
@@ -160,53 +168,54 @@ public static String filterTimes(String startTime, String endTime, JsonArray par
...
+                if (noAbsenceReason) {
+                    query += " OR reason_id IS NULL";
+                }
+                query += ")";
+            }
+            else if (noAbsenceReason) {
+                query += " AND reason_id IS NULL";
             }
+
+            query += (regularized != null ? " AND counsellor_regularisation = " + regularized + "" : "");
```
Comment (qpic-cgi) [inline presences/src/main/java/fr/openent/presences/helper/EventQueryHelper.java]: regularized + "" == regularized

File: presences/src/main/resources/public/ts/services/ReasonService.ts [hunk 12]
```diff
@@ -15,7 +16,7 @@ export interface ReasonService {
 export const reasonService: ReasonService = {
     getReasons: async (structureId: string, reasonTypeId?: Number): Promise<Reason[]> => {
         try {
-            let reasonTypeIdParam: string = reasonTypeId ? `&reasonTypeId=${reasonTypeId}` : '';
+            let reasonTypeIdParam: string = reasonTypeId != null && reasonTypeId != undefined ? `&reasonTypeId=${reasonTypeId}` : '';
```
Comment (qpic-cgi) [inline presences/src/main/resources/public/ts/services/ReasonService.ts:19]: !!reasonTypeId  sinon

File: presences/src/main/resources/public/ts/sniplets/events-form.ts [hunk 13]
```diff
@@ -399,7 +405,7 @@ const vm: ViewModel = {
     selectReason(): void {
         vm.selectedReason = vm.reasons.find(reason => reason.id === vm.form.reason_id);
         vm.canRegularize = (vm.selectedReason) ? (!vm.selectedReason.proving) : false;
-        vm.updateAbsenceRegularisation = vm.selectedReason ? vm.selectedReason.proving : false;
+        vm.updateAbsenceRegularisation = vm.selectedEventType == "ABSENCE" && (vm.selectedReason ? vm.selectedReason.proving : false);
```
Comment (qpic-cgi) [inline presences/src/main/resources/public/ts/sniplets/events-form.ts]: on a pas une constante ou un enum ?

File: presences/src/main/resources/public/ts/services/ReasonService.ts (hunk 12 above)
Comment (BastienBenard) [reply presences/src/main/resources/public/ts/services/ReasonService.ts:19]: Ne marche pas si reasonTypeId == 0

Comment (qpic-cgi) [reply presences/src/main/resources/public/ts/services/ReasonService.ts:19]: reasonTypeId >= 0

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 14]
```diff
@@ -959,4 +1007,12 @@ export const registersController = ng.controller('RegistersController',
                     window.filter = {};
                 }
             };
+
+            vm.getAbsenceReason = (): Reason[] => {
+                return vm.reasons.filter((reason: Reason) => reason.reason_type_id == REASON_TYPE_ID.ABSENCE)
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: point virgule à la fin

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 15]
```diff
@@ -959,4 +1007,12 @@ export const registersController = ng.controller('RegistersController',
...
+            };
+
+            vm.getLatenessReason = (): Reason[] => {
+                return vm.reasons.filter((reason: Reason) => reason.reason_type_id == REASON_TYPE_ID.LATENESS)
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: point virgule à la fin

File: presences/src/main/resources/public/ts/sniplets/events-form.ts [hunk 16]
```diff
@@ -641,6 +651,7 @@ const vm: ViewModel = {
         vm.eventBody.comment = vm.form.comment ? vm.form.comment : "";
         vm.eventBody.register_id = vm.form.register_id ? vm.form.register_id : -1;
         vm.eventBody.type_id = EventType.LATENESS;
+        vm.eventBody.reason_id = vm.form.reason_id
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/sniplets/events-form.ts]: point virgule à la fin

File: presences/src/main/resources/public/ts/sniplets/events-form.ts [hunk 17]
```diff
@@ -188,12 +194,14 @@ const vm: ViewModel = {
...
+            case EVENT_TYPE.ABSENCE:
+                vm.selectedEventType = EVENT_TYPE.ABSENCE;
+                vm.form.reason_id = vm.getAbsenceReason().find((reason: Reason) => reason.id == vm.form.reason_id) ? vm.form.reason_id : null
                 break;
-            case 'LATENESS':
-                vm.selectedEventType = 'LATENESS';
+            case EVENT_TYPE.LATENESS:
+                vm.selectedEventType = EVENT_TYPE.LATENESS;
                 vm.form.endDateTime = moment(new Date()).set({second: 0, millisecond: 0}).toDate();
+                vm.form.reason_id = vm.getLatenessReason().find((reason: Reason) => reason.id == vm.form.reason_id) ? vm.form.reason_id : null
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/sniplets/events-form.ts]: point virgule à la fin

File: presences/src/main/resources/public/ts/sniplets/events-form.ts [hunk 18]
```diff
@@ -188,12 +194,14 @@ const vm: ViewModel = {
 
     switchEventTypeForm: (eventType: TEventType): void => {
         switch (eventType) {
-            case 'ABSENCE':
-                vm.selectedEventType = 'ABSENCE';
+            case EVENT_TYPE.ABSENCE:
+                vm.selectedEventType = EVENT_TYPE.ABSENCE;
+                vm.form.reason_id = vm.getAbsenceReason().find((reason: Reason) => reason.id == vm.form.reason_id) ? vm.form.reason_id : null
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/sniplets/events-form.ts]: point virgule à la fin

File: presences/src/main/resources/public/ts/controllers/registers.ts [hunk 19]
```diff
@@ -713,6 +728,39 @@ export const registersController = ng.controller('RegistersController',
...
+                        events.register_id = parseInt(events.register_id);
+                    }
+                    eventService.updateEvent(events.id, {
+                        end_date: events.end_date,
+                        reason_id: events.reason_id,
+                        register_id: events.register_id,
+                        start_date: events.start_date,
+                        student_id: student.id,
+                        type_id: REASON_TYPE_ID.LATENESS
+                    })
```
Comment (JinDouang) [inline presences/src/main/resources/public/ts/controllers/registers.ts]: point virgule à la fin

**Review by JinDouang — APPROVED** [review]

**Pull request merged**